### PR TITLE
#839 doses to customer invoice

### LINKED ISSUE #839: [Vaccine Module][Custom Invoice][Doses] - Add doses to customer invoice
As per epic: https://github.com/sussol/org-issues/issues/23

#### Description

Need to add new column ('doses used' or 'doses given') in Customer Invoice. This column should only be rendered when there is an item that is a vaccine in the invoice. 

If at this stage item does not already have a getter `isVaccine`, create one, it should `safely` check for item.categroy.name = 'Vaccine'.

Maximum doses that can be entered is quantity * item.doses.
Minimum doses should be quantity.
If item.doses = 1, then the cell should not be editable and doses should automatically be set to quantity.

What if quantity is reduces and doses were entered ?
What if quantity is increased and doses were entered ?
I think for now, we just reset the doses to Minimum doses as per above ?

Design:

VM - Customer Invoices (1)-1

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #839 

Requirements from issue:
- Need to add new column ('doses used' or 'doses given') in Customer Invoice. This column should only be rendered when there is an item that is a vaccine in the invoice
    - ComponentDidMount - `Transaction.hasVaccine`
- If at this stage item does not already have a getter isVaccine, create one, it should safely check for item.categroy.name = 'Vaccine'.
    - Think a few setters might conflict with #849 
- Maximum doses that can be entered is quantity * item.doses.
- Minimum doses should be quantity.
- If item.doses = 1, then the cell should not be editable and doses should automatically be set to quantity.
    - Using `TransactionItem.setDoses(x)` and `TransactionItem.getValidDosesValue(x)` for ensuring valid quantities
- What if quantity is reduces and doses were entered / What if quantity is increased and doses were entered ?
    - Resetting doses to 0 in this case

### CODE REVIEW

File: src/pages/CustomerInvoicePage.js [hunk 1]
```diff
@@ -82,30 +129,39 @@ export class CustomerInvoicePage extends GenericPage {
...
   onAddMasterItems = () => {
-    this.props.runWithLoadingIndicator(() => {
-      this.props.database.write(() => {
-        this.props.transaction.addItemsFromMasterList(this.props.database);
-        this.props.database.save('Transaction', this.props.transaction);
+    const { database, transaction, runWithLoadingIndicator } = this.props;
+    runWithLoadingIndicator(() => {
+      database.write(() => {
+        transaction.addItemsFromMasterList(database);
+        database.save('Transaction', transaction);
```
Comment (andreievg) [inline src/pages/CustomerInvoicePage.js]: do we nee to check for vaccine added from master list here ?

File: src/pages/CustomerInvoicePage.js [hunk 2]
```diff
@@ -201,23 +257,54 @@ export class CustomerInvoicePage extends GenericPage {
...
+          isDisabled: isFinalised,
         };
+      case 'doses': {
+        // If not a vaccine item, render an empty cell.
+        // If a vaccine, display the doses of the item. Only
+        // editable if doses > 1.
+        const { item } = transactionItem;
+        const { isVaccine, doses } = item;
+        const hasDosesOfOne = doses === 1;
+        if (!isVaccine) return emptyCell;
```
Comment (andreievg) [inline src/pages/CustomerInvoicePage.js]: if doses is 0 or null should we also make it non enterable and empty ?

File: src/database/DataTypes/TransactionItem.js [hunk 3]
```diff
@@ -250,6 +250,72 @@ export class TransactionItem extends Realm.Object {
...
+   * @param {Number} value the number of doses to set
+   */
+  setDoses(value) {
+    const { doses: numberOfDosesPerVial } = this.item;
+    let dosesToAdd = this.getValidDosesValue(value);
+    const fifoBatches = this.batches.sorted('expiryDate');
+
+    fifoBatches.some(batch => {
+      const { totalQuantity: thisBatchTotalQuantity } = batch;
+      const maxDosesCanAddToBatch = thisBatchTotalQuantity * numberOfDosesPerVial;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionItem.js]: done 8f7052caeca5079d346bfc8d4d3ebd7913370a3e and fixed a variable here 7f48101adc8e3ccec16076fb70754e718cad6d5e - I found taking the floor of the whole equation to give a more accurate total number of doses applied.

File: src/database/DataTypes/TransactionItem.js [hunk 4]
```diff
@@ -250,6 +250,72 @@ export class TransactionItem extends Realm.Object {
...
+    });
+  }
+
+  /**
+   * Gets the total number of doses given for
+   * TransactionBatch's associated to this
+   * TransactionItem.
+   */
+  get totalDoses() {
+    const { batches } = this;
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionItem.js]: Cheers, done a2544907fb48ed75a17bab0b139b7dc71cb407b6

File: src/database/DataTypes/TransactionItem.js [hunk 5]
```diff
@@ -250,6 +250,72 @@ export class TransactionItem extends Realm.Object {
...
+    let dosesToAdd = this.getValidDosesValue(value);
+    const fifoBatches = this.batches.sorted('expiryDate');
+
+    fifoBatches.some(batch => {
+      const { totalQuantity: thisBatchTotalQuantity } = batch;
+      const maxDosesCanAddToBatch = thisBatchTotalQuantity * numberOfDosesPerVial;
+      const dosesAddingToThisBatch = Math.min(dosesToAdd, maxDosesCanAddToBatch);
+      batch.doses = dosesAddingToThisBatch;
+      dosesToAdd -= dosesAddingToThisBatch;
+      return !(dosesToAdd >= 0);
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/database/DataTypes/TransactionItem.js]: done 827f7437dcff1677f7cff351f7f06a412cd2ed24

File: src/pages/CustomerInvoicePage.js (hunk 2 above)
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: done a2ea237729812a9de505ac87f1d62ba68b8f8a6b - just going to treat a vaccine with doses nto set correctly as not a vaccine, and rendering an empty cell for doses?

File: src/pages/CustomerInvoicePage.js (hunk 1 above)
Comment (joshxg) [reply src/pages/CustomerInvoicePage.js]: Yes, I think consolidating as I have missed a few spots is a good idea. Also ensuring that to be a vaccine doses must be set correctly (as per comment below) is also a good idea. I think in a new issue? Happy to do it on this if that's a better idea

**Pull request merged**